File: 5words.cpp

#include <algorithm>
#include <chrono>
#include <cstring>
#include <fstream>
#include <iostream>
#include <list>
#include <map>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

//#define DO_PRINT   // print the results or not to save time
//#define PRINT_ALL  // printing the full 831 vs 538
#define PRINT_DEBUG       //
#define letter_buckets 6  // a cache size 0-26, but somewhere around 4-7 is best

#define bucket_shift (26 - letter_buckets)
#define total_buckets (1 << letter_buckets)
map<char, int> letter2mask;
map<int, char> mask2letter;
map<int, vector<vector<int>>> lowhighs;
#ifdef PRINT_ALL
unordered_map<int, vector<string>> mask2words;
#else
map<int, string> mask2word;
#endif
vector<vector<int>> bucket_matcher(total_buckets);

typedef std::chrono::high_resolution_clock::time_point TimeVar;
#define duration(a) \
  std::chrono::duration_cast<std::chrono::nanoseconds>(a).count()
#define timeNow() std::chrono::high_resolution_clock::now()
void printTime(string label, TimeVar t1) {
#ifdef PRINT_DEBUG
  cout << label << " " << duration(timeNow() - t1) / 1e6 << " msec\n";
#endif
}

int cnt = 0, cnt2 = 0;
int used[5];
void srch(int mask, char single, int depth) {
  if (depth == 5) {
    cnt += 1;

    if (!single) {
      single = mask2letter[(mask + 1) & (~mask)];
    }

#ifdef DO_PRINT
#ifdef PRINT_ALL
    for (auto word0 : mask2words[used[0]]) {
      for (auto word1 : mask2words[used[1]]) {
        for (auto word2 : mask2words[used[2]]) {
          for (auto word3 : mask2words[used[3]]) {
            for (auto word4 : mask2words[used[4]]) {
              cnt2 += 1;
              // printf("%s %s %s %s %s %c\n", word0.c_str(), word1.c_str(),
              //      word2.c_str(), word3.c_str(), word4.c_str(), single);
              cout << word0 << " " << word1 << " " << word2 << " " << word3
                   << " " << word4 << " " << single << endl;
            }
          }
        }
      }
    }
#else
    cout << mask2word[used[0]] << " " << mask2word[used[1]] << " "
         << mask2word[used[2]] << " " << mask2word[used[3]] << " "
         << mask2word[used[4]] << " " << single << endl;
#endif
#endif

    return;
  }

  while (true) {
    int first = (mask + 1) & (~mask);
    int highs = mask >> bucket_shift;
    // for (int h = 0; h < total_buckets; ++h) {
    // if (highs & h) {
    // continue;
    //}
    for (auto h : bucket_matcher[highs]) {
      for (auto m : lowhighs[first][h]) {
        if (mask & m) {
          continue;
        }
        used[depth] = m;
        srch(mask | m, single, depth + 1);
      }
    }

    if (single) {
      return;
    }
    single = mask2letter[first];
    mask |= first;
  }
}

int main() {
  TimeVar t1 = timeNow();

  // bucket_matcher cache?
  for (int high = 0; high < total_buckets; ++high) {
    for (int h = 0; h < total_buckets; ++h) {
      if ((h & high) == 0) {
        bucket_matcher[high].push_back(h);
      }
    }
  }

  int letter_rare[26];
  for (int i = 0; i < 26; ++i) {
    letter_rare[i] = 'a' + i;
  }
  vector<string> allwords;
  string filename = "words_alpha.txt";
  {
    ifstream file(filename, ios::binary | ios::ate);
    streamsize size = file.tellg();
    file.seekg(0, ios::beg);
    vector<char> buffer(size);
    if (file.read(buffer.data(), size)) {
      int linestart = 0;
      while (linestart < size) {
        int dupMask = 0;
        bool finishLine = false;
        for (int i = 0; i < 6; ++i) {
          char c = buffer[linestart + i];

          if (c == '\n') {
            if (i == 5) {
              // we have 5 letter word
              string word(buffer.begin() + linestart,
                          buffer.begin() + linestart + 5);
              // cout << linestart << word << endl;
              allwords.push_back(word);

              for (char& c : word) {
                letter_rare[c - 'a'] += 1 << 8;
              }
              linestart += 6;
              break;
            } else {
              // less than 5 letter word
              linestart += i + 1;
              break;
            }
          } else {
            if (i == 5) {
              // more than 5 letter word
              finishLine = true;
              linestart += 6;
              break;
            } else {
              // idk, still reading
              int ii = 1 << (c - 'a');
              if (dupMask & ii) {
                linestart += i + 1;
                finishLine = true;
                break;
              }
              dupMask |= ii;
            }
          }
        }
        if (finishLine) {
          while (buffer[linestart++] != '\n') {
          }
        }
      }
    }
  }

  sort(begin(letter_rare), end(letter_rare));
  for (int i = 0; i < 26; ++i) {
    int lr = letter_rare[i];
    char letter = lr & 0xff;
    int ii = 1 << i;

    letter2mask[letter] = ii;
    mask2letter[ii] = letter;
    lowhighs[ii] = vector<vector<int>>(total_buckets);
  }
  printTime("read words", t1);

  for (auto word : allwords) {
    int mask = 0;
    for (auto letter : word) {
      mask |= letter2mask[letter];
    }

#ifdef PRINT_ALL
    if (mask2words.count(mask)) {
      mask2words[mask].push_back(word);
    } else {
      mask2words[mask] = {word};
#else
    if (!mask2word.count(mask)) {
      mask2word[mask] = word;
#endif

      int low = (~(mask - 1)) & mask;    // first set bit
      int highs = mask >> bucket_shift;  // top n bits
      lowhighs[low][highs].push_back(mask);
    }
  }

#ifdef PRINT_DEBUG
  printf("words: %d   anagrams: %d\n", allwords.size(),
#ifdef PRINT_ALL
         mask2words.size()
#else
         mask2word.size()
#endif
  );
#endif
  printTime("preprocess", t1);

  srch(0, 0, 0);

#ifdef PRINT_DEBUG
  printf("%d %d %d \n", cnt, cnt2);
#endif
  printTime("total", t1);

  return 0;
}

File: 5words.py
from time import time
import itertools

start = time()

# pick a word list
filename = 'words_alpha.txt'
#filename = 'valid-wordle-words.txt'


ascii_offset = ord('a')  # cache this value for speed
allwords = []
letter_rare = [[0, chr(ascii_offset+i)] for i in range(26)]
with open(filename) as file:
    for line in file:
        # normally I'd do line.strip(), but that's slower
        # new line is the 6th letter
        if len(line) != 6:  # 5 letter words
            continue
        if len(set(line)) != 6:  # 5 unique letters
            continue
        word = line[:5]

        # for now we just store each word and count letter frequency
        allwords.append(word)
        for letter in word:
            letter_rare[ord(letter)-ascii_offset][0] += 1

# now that we have letter frequency, we'll be sorting each word based on
#  * it's rarest letter
#  * if it has the 5 most common letters (5 was picked experimentally)

# looking at top 5 common letters.
# caching some commonly used numbers/tables for it
letter_buckets = 5
bucket_shift = 26-letter_buckets
total_buckets = 1 << letter_buckets
bucket_matcher = [[h for h in range(total_buckets) if h & high == 0]
                  for high in range(total_buckets)]

# sort letters and assign masks based on rarity
letter_rare.sort()
letter2mask = dict()
mask2letter = dict()
lowhighs = dict()
for i, v in enumerate(letter_rare):
    letter = v[1]
    ii = 1 << i

    letter2mask[letter] = ii
    mask2letter[ii] = letter
    lowhighs[ii] = [[] for i in range(total_buckets)]

print('read words', time()-start)


# build a bitmask for each word
# and store the word in a bucket based on it's first bit and top few bits
mask2word = dict()
for word in allwords:
    mask = 0
    for letter in word:
        mask |= letter2mask[letter]

    if mask in mask2word:
        mask2word[mask].append(word)
    else:
        mask2word[mask] = [word]

        low = (~(mask-1)) & mask  # first set bit
        highs = mask >> bucket_shift  # top n bits
        lowhighs[low][highs].append(mask)

print('words:', len(allwords), 'anagrams:', len(mask2word))
print('preprocess', time()-start)

# for i in range(26):
#    print(i, letter_rare[i])
# sys.exit()


cnt = 0
cnt2 = 0
cnt3 = 0
used = [0]*5


def search(mask=0, single=False, depth=0):
    global cnt, cnt2, cnt3, start
    # cnt3[depth]+=1
    cnt3 += 1
    if depth == 5:
        # yay solution
        cnt += 1

        if not single:
            # find the missing letter if we haven't skipped one already
            single = mask2letter[(mask+1) & (~mask)]

        anagrams = [mask2word[u] for u in used]
        for word5 in itertools.product(*anagrams):
            cnt2 += 1
            print(*word5, single)
            #print('{:.3f}'.format(time()-start), *word5, single)
        return

    while True:
        # find the first unset bit
        first = (mask+1) & (~mask)
        highs = mask >> bucket_shift
        # try on that letter
        for h in bucket_matcher[highs]:
            for m in lowhighs[first][h]:
                if mask & m:
                    continue
                used[depth] = m
                search(mask | m, single, depth+1)

        # if we've already used our freebie, return
        if single:
            return
        # mark this letter as our freebie
        single = mask2letter[first]
        mask |= first


search()
print('{:.3f}'.format(time()-start), cnt, cnt2, cnt3)
